File: packages/static-generator/rust/compare.py
#!/usr/bin/env python3
"""Compare two native binaries on rendered HTML (Linux, one CPU, 1,000 documents)."""
import hashlib
import json
import os
from pathlib import Path
import random
import subprocess
import sys
import time

if len(sys.argv) < 4:
    raise SystemExit("Usage: compare.py baseline-binary candidate-binary page.html [another.html ...]")

os.sched_setaffinity(0, {min(os.sched_getaffinity(0))})
engines = {"before": sys.argv[1], "candidate": sys.argv[2]}
pages = [Path(path).read_text() for path in sys.argv[3:]]
request = (json.dumps({"version": 1, "id": 1, "operation": "minify_html",
                      "documents": [pages[i % len(pages)] for i in range(1000)]},
                     ensure_ascii=False) + "\n").encode()
report = {
    "methodology": "One CPU, 1000 documents repeating supplied HTML; 9 interleaved samples. "
                   "Wall time includes startup, stdin, native work and stdout collection. "
                   "No PHP adapter. Every output compared byte for byte.",
    "binary_sha256": {key: hashlib.sha256(Path(path).read_bytes()).hexdigest()
                      for key, path in engines.items()},
    "input_sha256": [hashlib.sha256(page.encode()).hexdigest() for page in pages],
    "request_bytes": len(request),
    "samples": [],
}
rng = random.Random(290912)
expected = None
for repeat in range(9):
    modes = list(engines)
    rng.shuffle(modes)
    for mode in modes:
        start = time.perf_counter()
        result = subprocess.run([engines[mode]], input=request, capture_output=True,
                                check=True, timeout=30)
        elapsed = (time.perf_counter() - start) * 1000
        if expected is None:
            response = json.loads(result.stdout)
            if (not isinstance(response, dict) or response.get("version") != 1
                    or response.get("id") != 1
                    or not isinstance(response.get("documents"), list)
                    or len(response["documents"]) != 1000
                    or not all(isinstance(document, str) for document in response["documents"])):
                raise RuntimeError("The first binary did not return a valid minification response")
            expected = result.stdout
        if result.stdout != expected:
            raise RuntimeError("Baseline and candidate outputs differ")
        report["samples"].append({"mode": mode, "repeat": repeat, "ms": elapsed})
print(json.dumps(report, indent=2))

File: packages/static-generator/rust/compare.php
<?php

declare(strict_types=1);

use Symfony\Component\Process\Process;

require __DIR__.'/../../../vendor/autoload.php';

$rawArguments = $_SERVER['argv'] ?? [];
if (! is_array($rawArguments)) {
    throw new InvalidArgumentException('Invalid command-line arguments.');
}

$arguments = [];
foreach ($rawArguments as $argument) {
    if (! is_string($argument)) {
        throw new InvalidArgumentException('Invalid command-line argument.');
    }

    $arguments[] = $argument;
}

if (count($arguments) < 4) {
    throw new InvalidArgumentException('Usage: php compare.php baseline-binary candidate-binary page.html [another.html ...]');
}

$status = file_get_contents('/proc/self/status');
if (false === $status || 1 !== preg_match('/^Cpus_allowed_list:\s*(\d+)/m', $status, $allowed)) {
    throw new RuntimeException('Cannot determine the first allowed Linux CPU.');
}

$affinity = new Process(['taskset', '-pc', $allowed[1], (string) getmypid()]);
$affinity->mustRun();

$binaries = ['before' => $arguments[1], 'candidate' => $arguments[2]];
$pages = array_map(static function (string $path): string {
    $html = file_get_contents($path);

    return false === $html ? throw new RuntimeException('Unreadable HTML file: '.$path) : $html;
}, array_slice($arguments, 3));
$documents = [];
for ($i = 0; $i < 1000; ++$i) {
    $documents[] = $pages[$i % count($pages)];
}

$encoded = array_map(static fn (string $html): string => json_encode($html, \JSON_UNESCAPED_UNICODE | \JSON_UNESCAPED_SLASHES | \JSON_THROW_ON_ERROR), $documents);
// Preserve the request bytes used by the earlier native-only measurements.
$request = '{"version": 1, "id": 1, "operation": "minify_html", "documents": ['.implode(', ', $encoded)."]}\n";
$report = [
    'methodology' => 'One CPU, 1000 documents repeating supplied HTML; 9 interleaved samples. Wall time includes startup, stdin, native work and stdout collection. No PHP adapter. Every output compared byte for byte.',
    'binary_sha256' => array_map(static fn (string $path): string => hash_file('sha256', $path) ?: throw new RuntimeException('Unreadable binary: '.$path), $binaries),
    'input_sha256' => array_map(static fn (string $html): string => hash('sha256', $html), $pages),
    'request_bytes' => strlen($request),
    'samples' => [],
];

mt_srand(290912);
$expected = null;
for ($repeat = 0; $repeat < 9; ++$repeat) {
    $modes = array_keys($binaries);
    shuffle($modes);
    foreach ($modes as $mode) {
        $process = new Process([$binaries[$mode]]);
        $process->setInput($request);
        $process->setTimeout(30);
        $start = hrtime(true);
        $process->mustRun();
        $elapsed = (hrtime(true) - $start) / 1e6;
        $output = $process->getOutput();
        if (null === $expected) {
            $response = json_decode($output, true, flags: \JSON_THROW_ON_ERROR);
            if (! is_array($response) || 1 !== ($response['version'] ?? null) || 1 !== ($response['id'] ?? null)
                || ! isset($response['documents']) || ! is_array($response['documents']) || 1000 !== count($response['documents'])
                || 1000 !== count(array_filter($response['documents'], is_string(...)))) {
                throw new RuntimeException('The first binary did not return a valid minification response.');
            }

            $expected = $output;
        }

        if ($output !== $expected) {
            throw new RuntimeException('Baseline and candidate outputs differ.');
        }

        $report['samples'][] = ['mode' => $mode, 'repeat' => $repeat, 'ms' => $elapsed];
    }
}

echo json_encode($report, \JSON_PRETTY_PRINT | \JSON_UNESCAPED_UNICODE | \JSON_UNESCAPED_SLASHES | \JSON_THROW_ON_ERROR)."\n";
